### PR TITLE
plannerd: improve lateral accuracy in turns

### PR DESCRIPTION
**Pull request by nuwandavek**
Example: 
image

More details in the report.

### CODE REVIEW

**Comment by sshane** [comment]
Had a couple of low-speed instances where it got into a sharp turn and either didn't straighten out, or took much too long to straighten out, despite the path visually near the center. Here's an instance where it wanted to do donuts 😆: https://connect.comma.ai/f15e3c37c118e841/1650438836550/1650438862469

**Comment by ntegan1** [comment]
definite improvement in low speed (20->40mph) turns, especially with torque control.

not sure if I could tell too much of a difference between 5e668b8 and a9952ea

File: selfdrive/controls/lib/lateral_planner.py [hunk 1]
```diff
@@ -42,8 +42,12 @@ def update(self, sm):
     self.LP.parse_model(md)
     if len(md.position.x) == TRAJECTORY_SIZE and len(md.orientation.x) == TRAJECTORY_SIZE:
       self.path_xyz = np.column_stack([md.position.x, md.position.y, md.position.z])
+      self.speed_forward = md.velocity.x
```
Comment (sshane) [inline selfdrive/controls/lib/lateral_planner.py:45]:
```suggestion
      self.speed_forward = np.array(md.velocity.x)
```

**Comment by nuwandavek** [comment]
> If plan.y.0 is from the device, shouldn't rotation arm be from the comma to the center of mass? (Negative) Edit: I probably have something backwards trying to look at it.

It should be that distance (+ve, as the sign is bakes into the model kinematic eqns). But we don't have that and `centerToFront` is a close approximation. But it doesn't do well in a bunch of cases. Rest of the changes in this PR should be a good improvement, and the rotation arm bit still needs some work.

File: selfdrive/controls/lib/lateral_planner.py [hunk 2]
```diff
@@ -59,26 +64,29 @@ def update(self, sm):
...
-    heading_pts = np.interp(v_ego * self.t_idxs[:LAT_MPC_N + 1], np.linalg.norm(self.path_xyz, axis=1), self.plan_yaw)
+      heading_cost = interp(v_ego, [0.0, 5.0, 10.0], [MPC_COST_LAT.HEADING2, MPC_COST_LAT.HEADING, MPC_COST_LAT.HEADING_LL])
+      self.lat_mpc.set_weights(MPC_COST_LAT.PATH, heading_cost, MPC_COST_LAT.LAT_JERK)
+
+    distance_forward = ((self.speed_forward[1:] + self.speed_forward[:-1]) / 2) * (self.t_idxs[1:] - self.t_idxs[:-1])
+    distance_forward = np.cumsum(np.insert(distance_forward, 0, 0))
+    y_pts = np.interp(v_ego * self.t_idxs[:LAT_MPC_N + 1], distance_forward, d_path_xyz[:, 1])
+    heading_pts = np.interp(v_ego * self.t_idxs[:LAT_MPC_N + 1], distance_forward, self.plan_yaw)
+    jerk_pts = np.interp(v_ego * self.t_idxs[:LAT_MPC_N + 1], distance_forward, self.lateral_jerk)
     self.y_pts = y_pts
```
Comment (ClockeNessMnstr) [inline selfdrive/controls/lib/lateral_planner.py:79]:
```suggestion
self.jerk_pts = jerk_pts

# Publish:
# lateralPlan.curvatureRates = [float(x) for x in self.lat_mpc.u_sol[0:CONTROL_N - 1]]  + [float(self.jerk_pts[CONTROL_N-1]/(self.speed_forward[CONTROL_N-1]**2))]
```

lateralPlan.curvatureRates[CONTROL_N] is not 0 anymore.

**Comment by ClockeNessMnstr** [comment]
what issues are happening with the rotation radius? are there some test case examples?

I had to "phase it in" from ~2.5m/s or so to avoid planner errors starting from a stop. but otherwise performance "seems fine" just testing from driving and logs.

File: selfdrive/controls/lib/lateral_mpc_lib/lat_mpc.py [hunk 3]
```diff
@@ -140,27 +144,29 @@ def reset(self, x0=np.zeros(X_DIM)):
...
-    self.yref[:,1] = heading_pts*(v_ego+5.0)
+    self.yref[:, 1] = heading_pts * (v_ego + 5.0)
+    self.yref[:, 2] = jerk_pts
     for i in range(N):
       self.solver.cost_set(i, "yref", self.yref[i])
-      self.solver.set(i, "p", p_cp)
-    self.solver.set(N, "p", p_cp)
-    self.solver.cost_set(N, "yref", self.yref[N][:2])
+      self.solver.set(i, "p", p_cp[i])
+    self.solver.set(N, "p", p_cp[1])
```
Comment (ClockeNessMnstr) [inline selfdrive/controls/lib/lateral_mpc_lib/lat_mpc.py:168]:
```suggestion
    self.solver.set(N, "p", p_cp[N])
```

File: selfdrive/controls/lib/lateral_mpc_lib/lat_mpc.py [hunk 4]
```diff
@@ -84,14 +86,16 @@ def gen_lat_ocp():
...
+  ocp.cost.yref_e = np.zeros((COST_DIM - 1, ))
   # TODO hacky weights to keep behavior the same
-  ocp.model.cost_y_expr = vertcat(y_ego,
-                                  ((v_ego +5.0) * psi_ego),
-                                  ((v_ego +5.0) * 4 * curv_rate))
-  ocp.model.cost_y_expr_e = vertcat(y_ego,
-                                    ((v_ego +5.0) * psi_ego))
+  costs = [
+    y_ego,
+    ((v_ego + 5.0) * psi_ego),
```
Comment (ClockeNessMnstr) [inline selfdrive/controls/lib/lateral_mpc_lib/lat_mpc.py:94]: # TODO hacky weights to keep behavior the same
Possible opportunity to get riid of this "hacky": ```+ 5.0``` ?

```suggestion
    (v_ego * psi_ego),
```

File: selfdrive/controls/lib/lateral_planner.py [hunk 5]
```diff
@@ -42,8 +44,12 @@ def update(self, sm):
     self.LP.parse_model(md)
     if len(md.position.x) == TRAJECTORY_SIZE and len(md.orientation.x) == TRAJECTORY_SIZE:
       self.path_xyz = np.column_stack([md.position.x, md.position.y, md.position.z])
+      self.speed_forward = np.linalg.norm(np.column_stack([md.velocity.x, md.velocity.y, md.velocity.z]), axis=1)
       self.t_idxs = np.array(md.position.t)
       self.plan_yaw = list(md.orientation.z)
+      self.plan_yaw_rate = list(md.orientationRate.z)
+      self.lateral_acc = self.speed_forward * self.plan_yaw_rate
+      self.jerk = np.gradient(self.lateral_acc, self.t_idxs)
```
Comment (ClockeNessMnstr) [inline selfdrive/controls/lib/lateral_planner.py:52]: should give a better approximation at t0. 
center points are already evaluated with second order approximation but default for the edges is 1st. 

```suggestion
      self.jerk = np.gradient(self.lateral_acc, self.t_idxs, edge_order=2)
```

**Comment by sshane** [comment]
More donuts with the model on master! https://connect.comma.ai/f15e3c37c118e841/1653552489341/1653552546691

**Comment by nuwandavek** [comment]
Status: This is stalled due to the distribution shifts in ground truth and model output distributions (especially velocity, when velocity is low). WIP a plan for using longitudinal planner predictions in the lateral planner to overcome the above mentioned issue. Running those experiments now.

**Comment by sshane** [comment]
This was merged in: https://github.com/commaai/openpilot/pull/25413

**Pull request closed without merging**